hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/checkpoint/KafkaConnectHdfsProvider.java

```diff
@@ -0,0 +1,140 @@
+    }
+
+    public boolean acceptDir(final Path path) {
+      final String dirName = path.getName();
+      final Matcher matcher = DIRECTORY_PATTERN.matcher(dirName);
+      return matcher.matches();
+    }
+  }
+
+  /**
+   * Convert map contains max offset of each partition to string.
+   * @param topic Topic name
+   * @param checkpoint Map with partition as key and max offset as value
+   * @return Checkpoint string
+   */
+  private static String buildCheckpointStr(final String topic,
+                                           final HashMap<Integer, Integer> checkpoint) {
+    final StringBuilder checkpointStr = new StringBuilder();
+    checkpointStr.append(topic);
+    for (int i = 0; i < checkpoint.size(); ++i) {
+      checkpointStr.append(",").append(i).append(":").append(checkpoint.get(i));
+    }
+    return checkpointStr.toString();
+  }
+
+  /**
+   * List file status recursively.
+   * @param curPath Current Path
+   * @param filter PathFilter
+   * @return All file status match kafka connect naming convention
+   * @throws IOException
+   */
+  private ArrayList<FileStatus> listAllFileStatus(Path curPath,
+                                                  KafkaConnectPathFilter filter) throws IOException {
+    ArrayList<FileStatus> allFileStatus = new ArrayList<>();
+    FileStatus[] fileStatus = this.fs.listStatus(curPath);
+    for (FileStatus status : fileStatus) {
+      if (status.isDirectory() && filter.acceptDir(status.getPath())) {
+        allFileStatus.addAll(listAllFileStatus(status.getPath(), filter));
+      } else {
+        if (filter.accept(status.getPath())) {
+          allFileStatus.add(status);
+        }
+      }
+    }
+    return allFileStatus;
+  }
+
+  @Override
+  public String getCheckpoint() throws IOException {
+    final KafkaConnectPathFilter filter = new KafkaConnectPathFilter();
+    final ArrayList<FileStatus> fileStatus = listAllFileStatus(this.path, filter);
+    if (fileStatus.size() == 0) {
+      throw new IOException("No valid Kafka Connect Hdfs file found under:" + this.path.getName());
+    }
+    final String topic = fileStatus.get(0).getPath().getName().split(FILENAME_SEPARATOR)[0];
+    int maxPartition = -1;
+    final HashMap<Integer, Integer> checkpointMap = new HashMap<>();
+    for (final FileStatus status : fileStatus) {
+      final String filename = status.getPath().getName();
+      final String[] groups = filename.split(FILENAME_SEPARATOR);
+      final int partition = Integer.parseInt(groups[1]);
+      final int offsetUpper = Integer.parseInt(groups[3]);
+      maxPartition = Math.max(maxPartition, partition);
+      if (checkpointMap.containsKey(partition)) {
+        checkpointMap.put(partition, Math.max(checkpointMap.get(partition), offsetUpper));
+      } else {
+        checkpointMap.put(partition, offsetUpper);
+      }
+    }
+    if (checkpointMap.size() != maxPartition + 1) {
+      throw new IOException("Missing partition from the file scan, "
```

Comment:
Let us add a test case to cover this scenario as well?

Reply:
sure